src/main/java/com/amazon/opendistroforelasticsearch/sql/executor/format/SelectResultSet.java

```diff
@@ -334,41 +348,67 @@ private String getFieldName(Field field) {
-                     * 3. If Type doesn't exist and isSelectAll() is true, Column creation for field is skipped
-                     */
-                    if (Schema.hasType(type)) {
-                        columns.add(
-                                new Schema.Column(
-                                        field,
-                                        fetchAlias(field, fieldMap),
-                                        Schema.Type.valueOf(type)
-                                )
-                        );
-                    } else if (!isSelectAll()) {
-                        throw new IllegalArgumentException(
-                                String.format("%s field types are currently not supported.", type));
-                    }
+            FieldMapping field = new FieldMapping(fieldName, typeMappings, fieldMap);
+            if (typeMappings.containsKey(fieldName) && !field.isMetaField()) {
+
+                if (field.isMultiField() && !field.isSpecified()) { continue; }
+                if (field.isPropertyField() && !field.isSpecified() && !field.isWildcardSpecified()) { continue; }
+
+                /*
+                 * Three cases regarding Type:
+                 * 1. If Type exists, create Column
+                 * 2. If Type doesn't exist and isSelectAll() is false, throw exception
+                 * 3. If Type doesn't exist and isSelectAll() is true, Column creation for fieldName is skipped
+                 */
+                String type = field.type().toUpperCase();
+                if (Schema.hasType(type)) {
+                    columns.add(
+                        new Schema.Column(
+                            fieldName,
+                            fetchAlias(fieldName, fieldMap),
+                            Schema.Type.valueOf(type)
+                        )
+                    );
+                } else if (!isSelectAll()) {
+                    throw new IllegalArgumentException(
+                        String.format("%s fieldName types are currently not supported.", type));
                 }
             }
         }
 
+        if (isSelectAllOnly(query)) {
+            populateAllNestedFields(columns, fieldNameList);
+        }
         return columns;
     }
 
+    /**
+     * SELECT * only without other columns or wildcard pattern specified.
+     */
+    private boolean isSelectAllOnly(Query query) {
+        return isSelectAll() && fetchFields(query).isEmpty();
+    }
+
+    /**
+     * Special case which trades off consistency of SELECT * meaning for more intuition from customer perspective.
+     * In other cases, * means all regular fields on the level.
+     * The only exception here is * picks all non-regular (nested) fields as JSON without flatten.
+     */
+    private void populateAllNestedFields(List<Schema.Column> columns, List<String> fields) {
+        Set<String> nestedFieldNames = fields.stream().
```

Comment:
That collection to set makes me suspicious. We're losing information here and doing it silently.
Should we fail fast? Add aliases? Or this will be handled elsewhere?

Reply:
Could you clarify what information is being lost? I don't understand what the issue is.

Reply:
nested field names is set, and we're flattening field mappings here, so in theory if there will be 2 fields with the same name, we will keep only one. If there are no chances that there will be duplicates, I'd expect to see List instead of Set

Reply:
> nested field names is set, and we're flattening field mappings here, so in theory if there will be 2 fields with the same name, we will keep only one. If there are no chances that there will be duplicates, I'd expect to see List instead of Set

I guess that's what this method meant to do. Will debug to confirm.

Reply:
> 
> 
> nested field names is set, and we're flattening field mappings here, so in theory if there will be 2 fields with the same name, we will keep only one. If there are no chances that there will be duplicates, I'd expect to see List instead of Set

I see, thanks for explanation. Good point.

Reply:
I just debugged to confirm. As I mentioned before, the culprit is still the ES API we're using. For SELECT *, It returns field list like "comment.likes", "comment.data", "message.author" etc. while we want to figure out all the nested field name like "comment", "message". That's why we're collecting their `path()` into Set. Does this make sense?

Reply:
I think in that case the name of the variable should not be `nestedFieldNames`, but rather `nestedFieldPaths` or `complexFields` or something similar.

Reply:
Sure, renaming to `nestedFieldPaths` for clarity.